### ISSUE TITLE
Left joins leads to nested queries in slick3.1.0-M2

### ISSUE THREAD

**Issue by mohittt8** [slick/slick#1241]
A query like 
       
    Compiled { (id: Rep[Int]) =>
           tableA
                 .joinLeft(tableB).on { case(tblA, tblB) => tblA.id === tblB.tblA_id} 
                 .filter { case(tblA, tblB) => tblA.id === id }
                  .map { case(tblA, tblB) => tblB}
generates a query like 
           
     SELECT `tblB fields` from tblB 
     LEFT OUTER JOIN ( SELECT `tblA fields` from tblA) on `join condition`
     WHERE `condition`. 

If possible, the query should be like 
           
     SELECT `tblB fields` from tblB 
     LEFT OUTER JOIN tblA  on `join condition`
     WHERE `condition`. 

The problem gets compounded if I have multiple left joins. That leads to several nested queries. 

I am using MySQL 5.6.

**Comment by szeiger** [comment]
The subquery starts with "select 1". This is used as a discriminator column to detect if the right-hand side of the join is empty. Simplifying this will be non-trivial. We'd have to pick a different discriminator from the already existing columns (i.e. any non-nullable column) and then convert to the required numeric type outside the join.

**Comment by szeiger** [comment]
Getting closer to reusing existing columns as discriminators. From

```
select s5.s62, s5.s63, s5.s64, s5.s65
from "A_NEWFUSION" s4
left outer join (
  select 1 as s62, "id" as s63, "a" as s64, "b" as s65
  from "A_NEWFUSION"
) s5
on s4."id" = s5.s63
where s4."id" = 3
```

to

```
select (case when (s5."id" is null) then null else 1 end), s5."id", s5."a", s5."b"
from "A_NEWFUSION" s4
left outer join "A_NEWFUSION" s5
on s4."id" = s5."id"
where s4."id" = 3
```

The discriminator conversion at the outer layer should be removable in `hoistClientOps` and the duplicate `id` columns would then be merged (as implemented in `hoistClientOps` since 3.1.0-M2).

I haven't looked at nested outer joins yet (other than running the tests to confirm that they still work).

**Comment by mohittt8** [comment]
@szeiger awesome :)
